controllers/nstemplateset/space_roles.go

```diff
@@ -0,0 +1,116 @@
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve last applied space roles")
+		}
+		// space roles to apply now
+		spaceRoleObjs, err := r.getSpaceRolesObjects(&ns, nsTmplSet.Spec.SpaceRoles)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve space roles to apply")
+		}
+
+		deleted, err := deleteObsoleteObjects(logger, r.Client, lastAppliedSpaceRoleObjs, spaceRoleObjs)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusUpdateFailed, err, "failed to delete redundant objects in namespace '%s'", ns.Name)
+		}
+		changed = changed || deleted
+		// labels to apply on all new objects
+		var labels = map[string]string{
+			toolchainv1alpha1.ProviderLabelKey: toolchainv1alpha1.ProviderLabelValue,
+			toolchainv1alpha1.OwnerLabelKey:    nsTmplSet.GetName(),
+		}
+		logger.Info("creating space role objects")
+		for _, obj := range spaceRoleObjs {
+			logger.Info(fmt.Sprintf("creating/updating %s/%s", obj.GetNamespace(), obj.GetName()))
+		}
+		// create (or update existing) objects based the tier template
+		applied, err := applycl.NewApplyClient(r.Client, r.Scheme).Apply(spaceRoleObjs, labels)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to provision namespace '%s' with space roles", ns.Name)
+		}
+		changed = changed || applied
+		// store the space roles in an annotation at the namespace level, so we know what was applied and how to deal with
+		// diffs when the space roles are changed (users added or removed, etc.)
+		sr, _ := json.Marshal(nsTmplSet.Spec.SpaceRoles)
+		if ns.Annotations == nil {
+			ns.Annotations = map[string]string{}
+		}
+		ns.Annotations[toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey] = string(sr)
+		if err := r.Client.Update(context.TODO(), &ns); err != nil {
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
+				"failed update namespace annotation")
+		}
+	}
+	return changed, nil
+}
+
+// Get the space role objects from the templates specified in the given `spaceRoles`
+// Returns the objects, or an error if something wrong happened when processing the templates
+func (r *spaceRolesManager) getSpaceRolesObjects(ns *corev1.Namespace, spaceRoles []toolchainv1alpha1.NSTemplateSetSpaceRole) ([]runtimeclient.Object, error) {
+	// store by kind and name
+	spaceRoleObjects := []runtimeclient.Object{}
+	for _, spaceRole := range spaceRoles {
+		tierTemplate, err := getTierTemplate(r.GetHostCluster, spaceRole.TemplateRef)
+		if err != nil {
+			return nil, err
+		}
+		for _, username := range spaceRole.Usernames {
+			objs, err := tierTemplate.process(r.Scheme, map[string]string{
+				Namespace: ns.Name,
+				Username:  username,
```

Comment:
btw, I just noticed that the `Username` parameter is used in all cases when the templates are processed - it could be moved to the `process` function itself. Can be done in separate PR :wink:

Reply:
I would rather keep it there for the sake of readability: having all params used to process a template in a single place is simpler for maintenance IMHO. Even if it causes a bit of duplication in the code